### PR TITLE
Add more detailed logging

### PR DESCRIPTION
**Pull request by lfdebrux**
### What problem does this pull request solve?

When looking at logs of failed end to end tests during deployments it can be hard to figure out where the tests successfully got to before failing.

This PR adds a bit more logging to the end to end tests, in a way that is configurable. There are now two extra environment variables that can be used to tweak the amount of logging done:

```bash
LOG_LEVEL=<debug|info|warn|error|fatal> TRACE=1 bundle exec rspec
```

There's a bunch more log statements in there as well (at the info level), with a sort of Gherkin like language, that's trying to describe the features we're trying to test as we're testing them. Hopefully this will help in following along with what's going on.

I stuck the tracing in because I recall someone mentioning it's hard to know what's succeeded in the tests, but it is a bit like trying to drink from a firehose, so if people think it's too much I'm happy to take it out. Something to note, I was a bit worried that the tracing would slow the tests down, however I haven't seen that in practice.

On the other hand, if people think we need more logging, I think this PR takes us close to the limit of what we can do with RSpec; I don't think it's really designed for these sorts of tests, which is why it doesn't have this functionality built in. If we find we need more insight into our tests still, I think we should consider moving to a different test framework.

<details><summary><h4>Example output with log level set to info</h4></summary>

```
% LOG_LEVEL=info GUI=1 SKIP_AUTH=1 FORMS_ADMIN_URL='http://localhost:3000/' bundle exec rspec

Full lifecycle of a form

Scenario: Form is created, made live by form admin user

As an editor user
When I create a new form
And make it live
Then I can share the live form
Scenario: Form is completed by a member of the public, and they answer all questions

As a form filler
When I fill out the new form
And I answer the questions that are not optional
Then I can check my answers before I submit them

As a form processor
When a form filler has submitted their answers
Then I can see their submission in my email inbox
And I can see their answers

Scenario: Form is completed by a member of the public, and they answer the questions that are not optional

As a form filler
When I fill out the new form
And I answer all of the questions
Then I can check my answers before I submit them

As a form processor
When a form filler has submitted their answers
Then I can see their submission in my email inbox
And I can see their answers

Scenario: Form is completed by a member of the public, and they request a confirmation email

As a form filler
When I fill out the new form
And I answer the questions that are not optional
Then I can check my answers before I submit them
And I can request a confirmation email

As a form processor
When a form filler has submitted their answers
Then I can see their submission in my email inbox
And I can see their answers

As a form filler
When I have filled out a form and requested a confirmation email
Then I can see the confirmation in my email inbox
  Form is created, made live by form admin user and completed by a member of the public

Finished in 37.37 seconds (files took 0.67566 seconds to load)
1 example, 0 failures

```

</details>

<details><summary><h4>Example output with all logging and tracing enabled</h4></summary>

```
% LOG_LEVEL=debug TRACE=1 GUI=1 SKIP_AUTH=1 FORMS_ADMIN_URL='http://localhost:3000/' bundle exec rspec

Full lifecycle of a form

Scenario: Form is created, made live by form admin user
TRACE:   features/complete_spec.rb:20: build_a_new_form

As an editor user
TRACE:  support/feature_helpers.rb:11: sign_in unless ENV.fetch("SKIP_AUTH", false)
TRACE:  support/feature_helpers.rb:12: visit '/'
TRACE:  support/feature_helpers.rb:14: expect(page).to have_content 'GOV.UK Forms'
TRACE:  support/feature_helpers.rb:16: delete_form
TRACE: support/feature_helpers.rb:186: visit forms_admin_url
TRACE:   support/feature_helpers.rb:5: ENV.fetch("FORMS_ADMIN_URL") { raise "You must set $FORMS_ADMIN_URL"}
TRACE: support/feature_helpers.rb:188: if page.has_link?(form_name)
TRACE:    features/complete_spec.rb:4: let(:form_name) { "capybara test form #{Time.now().strftime("%Y-%m-%d %H:%M.%S")}" }
When I create a new form
TRACE:  support/feature_helpers.rb:19: create_form_with_name(form_name)
TRACE:  support/feature_helpers.rb:77: click_link "Create a form"
TRACE:  support/feature_helpers.rb:78: expect(page.find("h1")).to have_content 'What’s the name of your form?'
TRACE:  support/feature_helpers.rb:79: fill_in "What’s the name of your form?", :with => form_name
TRACE:  support/feature_helpers.rb:80: click_button "Save and continue"
TRACE:  support/feature_helpers.rb:21: next_form_creation_step 'Add and edit your questions'
TRACE:  support/feature_helpers.rb:72: expect(page.find("h1")).to have_content 'Create a form'
TRACE:  support/feature_helpers.rb:73: click_link task
TRACE:  support/feature_helpers.rb:23: create_a_selection_question
TRACE:  support/feature_helpers.rb:84: expect(page.find("h1")).to have_content 'What kind of answer do you need to this question?'
TRACE:  support/feature_helpers.rb:85: choose "Selection from a list of options", visible: false
TRACE:  support/feature_helpers.rb:86: click_button "Continue"
TRACE:  support/feature_helpers.rb:88: expect(page.find("h1")).to have_content 'What’s your question?'
TRACE:  support/feature_helpers.rb:89: fill_in "What’s your question?", with: selection_question
TRACE:    features/complete_spec.rb:5: let(:selection_question) { "Do you want to remain anonymous?" }
TRACE:  support/feature_helpers.rb:90: click_button "Continue"
TRACE:  support/feature_helpers.rb:92: expect(page.find("h1")).to have_content 'Create a list of options'
TRACE:  support/feature_helpers.rb:93: check "People can only select one option", visible: false
TRACE:  support/feature_helpers.rb:94: fill_in "Option 1", :with => "Yes"
TRACE:  support/feature_helpers.rb:95: fill_in "Option 2", :with => "No"
TRACE:  support/feature_helpers.rb:96: click_button "Continue"
TRACE:  support/feature_helpers.rb:98: expect(page.find("h1")).to have_content 'Edit question'
TRACE:  support/feature_helpers.rb:99: click_button "Save question"
TRACE:  support/feature_helpers.rb:25: create_a_single_line_of_text_question
TRACE: support/feature_helpers.rb:103: within(page.find(".govuk-notification-banner__content")) do
TRACE: support/feature_helpers.rb:104: click_on "Add a question"
TRACE: support/feature_helpers.rb:107: expect(page.find("h1")).to have_content 'What kind of answer do you need to this question?'
TRACE: support/feature_helpers.rb:108: choose "Text", visible: false
TRACE: support/feature_helpers.rb:109: click_button "Continue"
TRACE: support/feature_helpers.rb:110: expect(page.find("h1")).to have_content 'How much text will people need to provide?'
TRACE: support/feature_helpers.rb:111: choose "Single line of text", visible: false
TRACE: support/feature_helpers.rb:112: click_button "Continue"
TRACE: support/feature_helpers.rb:113: expect(page.find("h1")).to have_content 'Edit question'
TRACE: support/feature_helpers.rb:114: fill_in "Question text", :with => question_text
TRACE:    features/complete_spec.rb:6: let(:question_text) { "What is your name?" }
TRACE: support/feature_helpers.rb:115: click_button "Save question"
TRACE:  support/feature_helpers.rb:27: first(:link, "your questions").click
TRACE:  support/feature_helpers.rb:29: add_a_route
TRACE: support/feature_helpers.rb:120: expect(page.find("h1")).to have_content 'Add and edit your questions'
TRACE: support/feature_helpers.rb:121: click_link "Add a question route"
TRACE: support/feature_helpers.rb:123: expect(page.find("h1")).to have_content 'Add a question route'
TRACE: support/feature_helpers.rb:124: choose "1. #{selection_question}", visible: false
TRACE: support/feature_helpers.rb:125: click_button "Continue"
TRACE: support/feature_helpers.rb:127: expect(page.find("h1")).to have_content 'Add a question route'
TRACE: support/feature_helpers.rb:128: select "Yes", from: "is answered as"
TRACE: support/feature_helpers.rb:129: select "Check your answers before submitting", from: "take the person to"
TRACE: support/feature_helpers.rb:130: click_button "Save and continue"
TRACE:  support/feature_helpers.rb:31: mark_pages_task_complete
TRACE: support/feature_helpers.rb:134: expect(page.find("h1")).to have_content 'Add and edit your questions'
TRACE: support/feature_helpers.rb:135: choose "Yes", visible: false
TRACE: support/feature_helpers.rb:136: click_button "Save and continue"
TRACE:  support/feature_helpers.rb:33: next_form_creation_step 'Add a declaration for people to agree to'
TRACE:  support/feature_helpers.rb:72: expect(page.find("h1")).to have_content 'Create a form'
TRACE:  support/feature_helpers.rb:73: click_link task
TRACE:  support/feature_helpers.rb:35: mark_declaration_task_complete
TRACE: support/feature_helpers.rb:140: expect(page.find("h1")).to have_content 'Add a declaration'
TRACE: support/feature_helpers.rb:141: choose "Yes", visible: false
TRACE: support/feature_helpers.rb:142: click_button "Save and continue"
TRACE:  support/feature_helpers.rb:37: next_form_creation_step 'Add information about what happens next'
TRACE:  support/feature_helpers.rb:72: expect(page.find("h1")).to have_content 'Create a form'
TRACE:  support/feature_helpers.rb:73: click_link task
TRACE:  support/feature_helpers.rb:39: expect(page.find("h1")).to have_content 'Add information about what happens next'
TRACE:  support/feature_helpers.rb:40: fill_in "Enter some information to tell people what will happen next", with: "We'll send you an email to let you know the outcome. You'll usually get a response within 10 working days."
TRACE:  support/feature_helpers.rb:41: click_button "Save and continue"
TRACE:  support/feature_helpers.rb:43: add_form_submission_email
TRACE: support/feature_helpers.rb:147: if page.has_content? "Enter the email address confirmation code"
TRACE: support/feature_helpers.rb:148: next_form_creation_step 'Set the email address completed forms will be sent to'
TRACE:  support/feature_helpers.rb:72: expect(page.find("h1")).to have_content 'Create a form'
TRACE:  support/feature_helpers.rb:73: click_link task
TRACE: support/feature_helpers.rb:150: expect(page.find("h1")).to have_content 'Set the email address for completed forms'
TRACE: support/feature_helpers.rb:152: expected_mail_reference = page.find('#notification-id', visible: false).value
TRACE: support/feature_helpers.rb:154: fill_in "What email address should completed forms be sent to?", with: test_email_address, fill_options: { clear: :backspace }
TRACE:    features/complete_spec.rb:2: let(:test_email_address) { "govuk-forms-automation-tests@digital.cabinet-office.gov.uk" }
TRACE: support/feature_helpers.rb:155: click_button "Save and continue"
TRACE: support/feature_helpers.rb:157: expect(page.find("h1")).to have_content 'Confirmation code sent'
TRACE: support/feature_helpers.rb:158: expect(page.find("main")).to have_content test_email_address
TRACE: support/feature_helpers.rb:160: click_link "Enter the email address confirmation code"
TRACE: support/feature_helpers.rb:162: expect(page.find("h1")).to have_content "Enter the confirmation code"
TRACE: support/feature_helpers.rb:164: confirmation_code = get_confirmation_from_notify(expected_mail_reference, confirmation_code: true)
TRACE: support/feature_helpers.rb:324: email = NotifyService.new.get_email(expected_mail_reference)
TRACE: support/feature_helpers.rb:326: start_time = Time.now
DEBUG: Waiting 3sec for mail delivery to do its thing.
TRACE: support/feature_helpers.rb:328: sleep 3
TRACE: support/feature_helpers.rb:329: try = 0
TRACE: support/feature_helpers.rb:330: while(Time.now - start_time < 5000) do
TRACE: support/feature_helpers.rb:330: while(Time.now - start_time < 5000) do
TRACE: support/feature_helpers.rb:331: try += 1
TRACE: support/feature_helpers.rb:333: if confirmation_code
TRACE: support/feature_helpers.rb:334: unless email.collection.first.body.nil?
TRACE: support/feature_helpers.rb:335: code = email.collection.first.body.match(/\d{6}/).to_s
DEBUG: Received the following code from Notify: “060525“
TRACE: support/feature_helpers.rb:337: return code
TRACE: support/feature_helpers.rb:166: abort("ABORT!!! #{expected_mail_reference} could not be found in Notify!!!") unless confirmation_code
TRACE: support/feature_helpers.rb:168: fill_in "Enter the confirmation code", with: confirmation_code
TRACE: support/feature_helpers.rb:170: click_button "Save and continue"
TRACE: support/feature_helpers.rb:172: expect(page.find("h1")).to have_content 'Email address confirmed'
TRACE: support/feature_helpers.rb:174: click_link "Continue creating a form"
TRACE:  support/feature_helpers.rb:45: next_form_creation_step 'Provide a link to privacy information for this form'
TRACE:  support/feature_helpers.rb:72: expect(page.find("h1")).to have_content 'Create a form'
TRACE:  support/feature_helpers.rb:73: click_link task
TRACE:  support/feature_helpers.rb:47: expect(page.find("h1")).to have_content 'Provide a link to privacy information for this form'
TRACE:  support/feature_helpers.rb:48: fill_in "Enter a link to privacy information for this form", with: "https://www.gov.uk/help/privacy-notice"
TRACE:  support/feature_helpers.rb:49: click_button "Save and continue"
TRACE:  support/feature_helpers.rb:51: next_form_creation_step 'Provide contact details for support'
TRACE:  support/feature_helpers.rb:72: expect(page.find("h1")).to have_content 'Create a form'
TRACE:  support/feature_helpers.rb:73: click_link task
TRACE:  support/feature_helpers.rb:53: expect(page.find("h1")).to have_content "Provide contact details for support"
TRACE:  support/feature_helpers.rb:54: check "Email", visible: false
TRACE:  support/feature_helpers.rb:55: fill_in "Enter the email address", with: test_email_address
TRACE:  support/feature_helpers.rb:56: click_button "Save and continue"
And make it live
TRACE:  support/feature_helpers.rb:59: next_form_creation_step 'Make your form live'
TRACE:  support/feature_helpers.rb:72: expect(page.find("h1")).to have_content 'Create a form'
TRACE:  support/feature_helpers.rb:73: click_link task
TRACE:  support/feature_helpers.rb:61: expect(page.find("h1")).to have_content "Make your form live"
TRACE:  support/feature_helpers.rb:62: choose "Yes", visible: false
TRACE:  support/feature_helpers.rb:63: click_button "Save and continue"
TRACE:  support/feature_helpers.rb:64: expect(page.find("h1")).to have_content "Your form is live"
TRACE:  support/feature_helpers.rb:66: click_link "Continue to form details"
TRACE:  support/feature_helpers.rb:68: expect(page.find("h1")).to have_content form_name
Then I can share the live form
TRACE:   features/complete_spec.rb:23: live_form_link = page.find('[data-copy-target]').text
TRACE:   features/complete_spec.rb:25: unless bypass_end_to_end_tests('forms-runner', live_form_link)
TRACE: support/feature_helpers.rb:355: visit link
TRACE: support/feature_helpers.rb:357: alert_message = if service_name == "forms-admin"
TRACE: support/feature_helpers.rb:359: elsif service_name == "forms-runner"
TRACE: support/feature_helpers.rb:360: "forms-runner is running in maintenance mode...aborting any further e2e tests forms-runner"
TRACE: support/feature_helpers.rb:363: return false unless current_path.end_with?("/maintenance")
Scenario: Form is completed by a member of the public, and they answer all questions
TRACE:   features/complete_spec.rb:28: form_is_filled_in_by_form_filler(live_form_link, skip_question: false)

As a form filler
When I fill out the new form
TRACE: support/feature_helpers.rb:210: visit live_form_link
TRACE: support/feature_helpers.rb:212: if skip_question
And I answer the questions that are not optional
TRACE: support/feature_helpers.rb:217: answer_selection_question("No")
TRACE: support/feature_helpers.rb:292: choose text, visible: false
TRACE: support/feature_helpers.rb:293: click_button "Continue"
TRACE: support/feature_helpers.rb:219: expect(page).to have_content question_text
TRACE: support/feature_helpers.rb:220: answer_single_line(answer_text)
TRACE:    features/complete_spec.rb:7: let(:answer_text) { "test name" }
TRACE: support/feature_helpers.rb:287: fill_in 'question[text]', with: text
TRACE: support/feature_helpers.rb:288: click_button 'Continue'
Then I can check my answers before I submit them
TRACE: support/feature_helpers.rb:224: expect(page).to have_content 'Check your answers before submitting your form'
TRACE: support/feature_helpers.rb:226: if skip_question
TRACE: support/feature_helpers.rb:230: expect(page).to have_content selection_question
TRACE: support/feature_helpers.rb:231: expect(page).to have_content "No"
TRACE: support/feature_helpers.rb:233: expect(page).to have_content answer_text
TRACE: support/feature_helpers.rb:236: expected_mail_reference = page.find('#notification-id', visible: false).value
TRACE: support/feature_helpers.rb:237: expected_confirmation_mail_reference = nil
TRACE: support/feature_helpers.rb:239: if page.has_content? "Do you want to get an email confirming your form has been submitted?"
TRACE: support/feature_helpers.rb:240: if confirmation_email
TRACE: support/feature_helpers.rb:246: choose "No", visible: false
TRACE: support/feature_helpers.rb:250: click_button 'Submit'
TRACE: support/feature_helpers.rb:252: expect(page).to have_content 'Your form has been submitted'

As a form processor
When a form filler has submitted their answers
Then I can see their submission in my email inbox
TRACE: support/feature_helpers.rb:258: form_submission_email = get_confirmation_from_notify(expected_mail_reference)
TRACE: support/feature_helpers.rb:324: email = NotifyService.new.get_email(expected_mail_reference)
TRACE: support/feature_helpers.rb:326: start_time = Time.now
DEBUG: Waiting 3sec for mail delivery to do its thing.
TRACE: support/feature_helpers.rb:328: sleep 3
TRACE: support/feature_helpers.rb:329: try = 0
TRACE: support/feature_helpers.rb:330: while(Time.now - start_time < 5000) do
TRACE: support/feature_helpers.rb:330: while(Time.now - start_time < 5000) do
TRACE: support/feature_helpers.rb:331: try += 1
TRACE: support/feature_helpers.rb:333: if confirmation_code
TRACE: support/feature_helpers.rb:340: unless email.collection.first.status.nil?
TRACE: support/feature_helpers.rb:341: status = email.collection.first.status
DEBUG: Received the following status from Notify: “delivered“
TRACE: support/feature_helpers.rb:343: return email.collection.first
TRACE: support/feature_helpers.rb:260: abort("ABORT!!! #{expected_mail_reference} could not be found in Notify!!!") unless form_submission_email
And I can see their answers
TRACE: support/feature_helpers.rb:263: if skip_question
TRACE: support/feature_helpers.rb:267: expect(form_submission_email.body).to have_content selection_question
TRACE: support/feature_helpers.rb:268: expect(form_submission_email.body).to have_content "No"
TRACE: support/feature_helpers.rb:270: expect(form_submission_email.body).to have_content question_text
TRACE: support/feature_helpers.rb:271: expect(form_submission_email.body).to have_content answer_text
TRACE: support/feature_helpers.rb:274: if expected_confirmation_mail_reference

Scenario: Form is completed by a member of the public, and they answer the questions that are not optional
TRACE:   features/complete_spec.rb:32: form_is_filled_in_by_form_filler(live_form_link, skip_question: true)

As a form filler
When I fill out the new form
TRACE: support/feature_helpers.rb:210: visit live_form_link
TRACE: support/feature_helpers.rb:212: if skip_question
And I answer all of the questions
TRACE: support/feature_helpers.rb:214: answer_selection_question("Yes")
TRACE: support/feature_helpers.rb:292: choose text, visible: false
TRACE: support/feature_helpers.rb:293: click_button "Continue"
Then I can check my answers before I submit them
TRACE: support/feature_helpers.rb:224: expect(page).to have_content 'Check your answers before submitting your form'
TRACE: support/feature_helpers.rb:226: if skip_question
TRACE: support/feature_helpers.rb:227: expect(page).to have_content selection_question
TRACE: support/feature_helpers.rb:228: expect(page).to have_content "Yes"
TRACE: support/feature_helpers.rb:236: expected_mail_reference = page.find('#notification-id', visible: false).value
TRACE: support/feature_helpers.rb:237: expected_confirmation_mail_reference = nil
TRACE: support/feature_helpers.rb:239: if page.has_content? "Do you want to get an email confirming your form has been submitted?"
TRACE: support/feature_helpers.rb:240: if confirmation_email
TRACE: support/feature_helpers.rb:246: choose "No", visible: false
TRACE: support/feature_helpers.rb:250: click_button 'Submit'
TRACE: support/feature_helpers.rb:252: expect(page).to have_content 'Your form has been submitted'

As a form processor
When a form filler has submitted their answers
Then I can see their submission in my email inbox
TRACE: support/feature_helpers.rb:258: form_submission_email = get_confirmation_from_notify(expected_mail_reference)
TRACE: support/feature_helpers.rb:324: email = NotifyService.new.get_email(expected_mail_reference)
TRACE: support/feature_helpers.rb:326: start_time = Time.now
DEBUG: Waiting 3sec for mail delivery to do its thing.
TRACE: support/feature_helpers.rb:328: sleep 3
TRACE: support/feature_helpers.rb:329: try = 0
TRACE: support/feature_helpers.rb:330: while(Time.now - start_time < 5000) do
TRACE: support/feature_helpers.rb:330: while(Time.now - start_time < 5000) do
TRACE: support/feature_helpers.rb:331: try += 1
TRACE: support/feature_helpers.rb:333: if confirmation_code
TRACE: support/feature_helpers.rb:340: unless email.collection.first.status.nil?
TRACE: support/feature_helpers.rb:341: status = email.collection.first.status
DEBUG: Received the following status from Notify: “delivered“
TRACE: support/feature_helpers.rb:343: return email.collection.first
TRACE: support/feature_helpers.rb:260: abort("ABORT!!! #{expected_mail_reference} could not be found in Notify!!!") unless form_submission_email
And I can see their answers
TRACE: support/feature_helpers.rb:263: if skip_question
TRACE: support/feature_helpers.rb:264: expect(form_submission_email.body).to have_content selection_question
TRACE: support/feature_helpers.rb:265: expect(form_submission_email.body).to have_content "Yes"
TRACE: support/feature_helpers.rb:274: if expected_confirmation_mail_reference

Scenario: Form is completed by a member of the public, and they request a confirmation email
TRACE:   features/complete_spec.rb:36: form_is_filled_in_by_form_filler(live_form_link, confirmation_email: test_email_address)

As a form filler
When I fill out the new form
TRACE: support/feature_helpers.rb:210: visit live_form_link
TRACE: support/feature_helpers.rb:212: if skip_question
And I answer the questions that are not optional
TRACE: support/feature_helpers.rb:217: answer_selection_question("No")
TRACE: support/feature_helpers.rb:292: choose text, visible: false
TRACE: support/feature_helpers.rb:293: click_button "Continue"
TRACE: support/feature_helpers.rb:219: expect(page).to have_content question_text
TRACE: support/feature_helpers.rb:220: answer_single_line(answer_text)
TRACE: support/feature_helpers.rb:287: fill_in 'question[text]', with: text
TRACE: support/feature_helpers.rb:288: click_button 'Continue'
Then I can check my answers before I submit them
TRACE: support/feature_helpers.rb:224: expect(page).to have_content 'Check your answers before submitting your form'
TRACE: support/feature_helpers.rb:226: if skip_question
TRACE: support/feature_helpers.rb:230: expect(page).to have_content selection_question
TRACE: support/feature_helpers.rb:231: expect(page).to have_content "No"
TRACE: support/feature_helpers.rb:233: expect(page).to have_content answer_text
TRACE: support/feature_helpers.rb:236: expected_mail_reference = page.find('#notification-id', visible: false).value
TRACE: support/feature_helpers.rb:237: expected_confirmation_mail_reference = nil
TRACE: support/feature_helpers.rb:239: if page.has_content? "Do you want to get an email confirming your form has been submitted?"
TRACE: support/feature_helpers.rb:240: if confirmation_email
And I can request a confirmation email
TRACE: support/feature_helpers.rb:242: choose "Yes", visible: false
TRACE: support/feature_helpers.rb:243: fill_in "email_confirmation_form[confirmation_email_address]", with: confirmation_email
TRACE: support/feature_helpers.rb:244: expected_confirmation_mail_reference = page.find("#confirmation-email-reference", visible: false).value
TRACE: support/feature_helpers.rb:250: click_button 'Submit'
TRACE: support/feature_helpers.rb:252: expect(page).to have_content 'Your form has been submitted'

As a form processor
When a form filler has submitted their answers
Then I can see their submission in my email inbox
TRACE: support/feature_helpers.rb:258: form_submission_email = get_confirmation_from_notify(expected_mail_reference)
TRACE: support/feature_helpers.rb:324: email = NotifyService.new.get_email(expected_mail_reference)
TRACE: support/feature_helpers.rb:326: start_time = Time.now
DEBUG: Waiting 3sec for mail delivery to do its thing.
TRACE: support/feature_helpers.rb:328: sleep 3
TRACE: support/feature_helpers.rb:329: try = 0
TRACE: support/feature_helpers.rb:330: while(Time.now - start_time < 5000) do
TRACE: support/feature_helpers.rb:330: while(Time.now - start_time < 5000) do
TRACE: support/feature_helpers.rb:331: try += 1
TRACE: support/feature_helpers.rb:333: if confirmation_code
TRACE: support/feature_helpers.rb:340: unless email.collection.first.status.nil?
TRACE: support/feature_helpers.rb:341: status = email.collection.first.status
DEBUG: Received the following status from Notify: “delivered“
TRACE: support/feature_helpers.rb:343: return email.collection.first
TRACE: support/feature_helpers.rb:260: abort("ABORT!!! #{expected_mail_reference} could not be found in Notify!!!") unless form_submission_email
And I can see their answers
TRACE: support/feature_helpers.rb:263: if skip_question
TRACE: support/feature_helpers.rb:267: expect(form_submission_email.body).to have_content selection_question
TRACE: support/feature_helpers.rb:268: expect(form_submission_email.body).to have_content "No"
TRACE: support/feature_helpers.rb:270: expect(form_submission_email.body).to have_content question_text
TRACE: support/feature_helpers.rb:271: expect(form_submission_email.body).to have_content answer_text
TRACE: support/feature_helpers.rb:274: if expected_confirmation_mail_reference

As a form filler
When I have filled out a form and requested a confirmation email
Then I can see the confirmation in my email inbox
TRACE: support/feature_helpers.rb:280: confirmation_email_notification = get_confirmation_from_notify(expected_confirmation_mail_reference)
TRACE: support/feature_helpers.rb:324: email = NotifyService.new.get_email(expected_mail_reference)
TRACE: support/feature_helpers.rb:326: start_time = Time.now
DEBUG: Waiting 3sec for mail delivery to do its thing.
TRACE: support/feature_helpers.rb:328: sleep 3
TRACE: support/feature_helpers.rb:329: try = 0
TRACE: support/feature_helpers.rb:330: while(Time.now - start_time < 5000) do
TRACE: support/feature_helpers.rb:330: while(Time.now - start_time < 5000) do
TRACE: support/feature_helpers.rb:331: try += 1
TRACE: support/feature_helpers.rb:333: if confirmation_code
TRACE: support/feature_helpers.rb:340: unless email.collection.first.status.nil?
TRACE: support/feature_helpers.rb:341: status = email.collection.first.status
DEBUG: Received the following status from Notify: “delivered“
TRACE: support/feature_helpers.rb:343: return email.collection.first
TRACE: support/feature_helpers.rb:282: abort("ABORT!!! #{expected_confirmation_mail_reference} could not be found in Notify!!!") unless confirmation_email_notification
TRACE:   features/complete_spec.rb:39: delete_form
TRACE: support/feature_helpers.rb:186: visit forms_admin_url
TRACE:   support/feature_helpers.rb:5: ENV.fetch("FORMS_ADMIN_URL") { raise "You must set $FORMS_ADMIN_URL"}
TRACE: support/feature_helpers.rb:188: if page.has_link?(form_name)
TRACE: support/feature_helpers.rb:189: click_link(form_name, match: :one)
TRACE: support/feature_helpers.rb:190: live_form_url = page.current_url
TRACE: support/feature_helpers.rb:191: delete_path = live_form_url.gsub("live", "delete")
TRACE: support/feature_helpers.rb:193: visit delete_path
TRACE: support/feature_helpers.rb:194: expect(page.find("h1")).to have_content 'Are you sure you want to delete this draft?'
TRACE: support/feature_helpers.rb:195: choose "Yes", visible: false
TRACE: support/feature_helpers.rb:196: click_button "Continue"
TRACE: support/feature_helpers.rb:197: expect(page.find("h1")).to have_content 'GOV.UK Forms'
TRACE: support/feature_helpers.rb:198: expect(page.find(".govuk-notification-banner")).to have_content "Successfully deleted ‘#{form_name}’"
TRACE: support/feature_helpers.rb:199: if page.has_css?('.govuk-table')
TRACE: support/feature_helpers.rb:200: expect(page.find('.govuk-table')).not_to have_content form_name
  Form is created, made live by form admin user and completed by a member of the public

Finished in 37.77 seconds (files took 0.7349 seconds to load)
1 example, 0 failures


```

</details>

### Things to consider when reviewing

- Ensure that you consider the wider context.
- Does it work when run on your machine?
- Is it clear what the code is doing?
- Do the commit messages explain why the changes were made?
- Has all relevant documentation been updated?

### CODE REVIEW

File: spec/support/feature_helpers.rb [hunk 1]
```diff
@@ -199,17 +203,24 @@ def delete_form
...
+    logger.info "As a form filler"
+
+    logger.info "When I fill out the new form"
     visit live_form_link
 
     if skip_question
+      logger.info "And I answer all of the questions"
       answer_selection_question("Yes")
     else
+      logger.info "And I answer the questions that are not optional"
```
Comment (DavidBiddle) [inline spec/support/feature_helpers.rb:216]: As above:
```suggestion
      logger.info "And I choose the answer with a route attached"
```

**Review by DavidBiddle — APPROVED** [review]
Nice - I like how the `info` statements are kind of doing double duty as comments/documentation. The gherkin-like syntax is nice too! Made a minor suggestion about the wording of the log messages around routing.

File: spec/features/complete_spec.rb [hunk 2]
```diff
@@ -11,16 +11,28 @@
...
+      logger.info("Then I can share the live form")
       live_form_link = page.find('[data-copy-target]').text
 
       unless bypass_end_to_end_tests('forms-runner', live_form_link)
         # Testing alternate routes (basic routing with a skip question)
+        logger.info "Scenario: Form is completed by a member of the public, and they answer all questions"
         form_is_filled_in_by_form_filler(live_form_link, skip_question: false)
+
+        logger.info
+        logger.info "Scenario: Form is completed by a member of the public, and they answer the questions that are not optional"
```
[the remark this answers is not in the document]
Comment (lfdebrux) [reply spec/features/complete_spec.rb:31]: Yes now that I think about it optional questions is a different feature to routing 🤦 

Thanks for the suggestions!

[the remark this answers is not in the document]
Comment (DavidBiddle) [reply spec/features/complete_spec.rb:31]: Wouldn't be a problem if I'd written a useful comment in the first place 😬

**Pull request merged**